sklearn/discriminant_analysis.py

```diff
@@ -470,65 +483,68 @@ def _solve_svd(self, X, y):
         y : array-like of shape (n_samples,) or (n_samples, n_targets)
             Target values.
         """
+        xp, is_array_api = get_namespace(X)
+
+        if is_array_api:
+            svd = xp.linalg.svd
+        else:
+            svd = scipy.linalg.svd
```

Comment:
I'm new to the array API and such, so there is a lot I don't know. But I'm interested to learn because I'd like to contribute to efforts like this.

When reading this PR to learn more I was expecting to mostly see `np.` be replaced with `xp.` and very little or no `if` statements that depend on what kind of array is being processed.

Why not use `svd = xp.linalg.svd` for numpy and "array API" inputs? Maybe because for `a = np.array(...)`, `a` doesn't have an `__array_namespace__`? Which made me wonder why it doesn't have that attribute. Now it feels like "one question leading to another and to another, ..." - so my ultimate question: what is a good place to talk about these things? (It feels like this PR is somehow the wrong place but I don't know a better one)

Reply:
> [...] what is a good place to talk about these things? (It feels like this PR is somehow the wrong place but I don't know a better one)

+1

The official organisation leading the Array API standard is [`data-apis`](https://github.com/data-apis/). [`data-apis/array-api`](https://github.com/data-apis/array-api) is an official repository versioning [the static website of the standard](https://data-apis.org/array-api/latest/) but apart from its issues tracker, I do not think there is a dedicated forum.

I think @scientific-python might have been a suitable organisation to host discussions (and the standard?) but it postdates the standard (IIRC) and probably the direction of both initiative is different. (Under @scientific-python, the Array API likely would have been a [SPEC](https://scientific-python.org/specs/) IMO.)

Reply:
I think https://github.com/data-apis/array-api/issues is a perfectly reasonable place to bring up these kinds of usability or "why was this choice made" discussions - and there are other such conversations there already. There's folks from many libraries participating there, so it's probably preferred over the issue tracker or mailing list of a single project. Currently issues are fine, maybe at some point the volume of such questions will grow that using a separate Discourse is helpful, but that is not the case today.

> I think @scientific-python might have been a suitable organisation to host discussions (and the standard?) but it postdates the standard (IIRC) and probably the direction of both initiative is different. (Under @scientific-python, the Array API likely would have been a [SPEC](https://scientific-python.org/specs/) IMO.)

They're complementary. I don't think the direction is different, just the scopes. The Data API Consortium is squarely focusing on array and dataframe APIs and interoperability features/protocols around those. Scientific Python has a "best practices for Python projects" in general scope - and not even limited to technical topics, also best practices around communities.

The whole API standard wouldn't have been a single SPEC, the scope and level of detail in that standard is way too much for a single SPEC. See a SPEC more like a PEP or a NEP, and the array API standard as a whole open source project.

Reply:
> Why not use `svd = xp.linalg.svd` for numpy and "array API" inputs?

I think the point here is that in this case, `scipy.linalg.svd` is preferred over `numpy.linalg.svd` (it may be faster), and `xp.linalg.svd` would select the `numpy` function.

I'll also note that indeed that `numpy.ndarray` does not yet have a `__array_namespace__` method as you noticed. The reason for that is that the main `numpy` namespace is not (yet) standard-compliant. Getting there takes time, due to backwards compatibility reasons. The new `numpy.array_api` namespace could be made compliant straight away. Over time, `numpy` aims to evolve its main namespace towards full compliance. The NumPy docs have [a table with differences](https://numpy.org/devdocs/reference/array_api.html?highlight=array_api#table-of-differences-between-numpy-array-api-and-numpy).

Reply:
For this specific PR, is the scipy SVD really better than numpy SVD? Do they use different LAPACK drivers by default? Based on the documentation it seems that both seems to use `gesdd` by default:

- https://numpy.org/doc/stable/reference/generated/numpy.linalg.svd.html
- https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.svd.html

If my analysis is wrong, and the choice of the LAPACK driver happens to be important (e.g. to trade of speed for numerical stability) shall we suggest an extension the Array API spec to allow for driver hints to favor one SVD algorithm of the other and fallback to an implementation-specific default otherwise?

Reply:
It's the same driver routine underneath indeed. So I think it should be the same - the differences are:
1. NumPy may have been built without a proper LAPACK dependency (using the f2c'd `lapack_lite` sources instead). This may still be fairly common when folks build from source.
2. NumPy using the CLAPACK interface, while SciPy uses the Fortran interface (this shouldn't matter all that much).